doc/design/state-machine.md

```diff
@@ -0,0 +1,249 @@
+Therefore, the state machines of multiple AGs are effectively bridged by the state machines of their overlapping cells.
+
+<p style="text-align: center;">
+  <img src="img/AG-state-machine.png" title="AG" alt="AG" width="70%"/>
+</p>
+
+### States
+
+__`Pending`__: the AG is waiting to be scheduled.
+
+__`Preempting`__: the AG has reserved cells but is waiting for the completion of preemptions of other AGs.
+
+__`Allocated`__: the AG is allocated cells.
+
+__`Being preempted`__: the AG is being preempted by other AGs (the preemption is still ongoing).
+
+__`Deleted`__: the AG is fully deleted and all of its cells are released.
+
+Note that only ``Pending``, `Allocated`, and `Deleted` are persistent, thus they are the recovery points of HiveD. While the other AG states (`Preempting`, `Being preempted`) are volatile, so they will transition to others after scheduler crash and restart (i.e., e<sub>c</sub> in the state machine).
+
+Also note that `Allocated` state includes updating pod annotation, pod binding, and pod running. We assume once pod annotation has been updated, the pod binding and pod running are handled by K8s. We hence only describe the cell allocation state in the state machine, and do not care about the pods' real running state.
+
+### Common AG life cycles
+
+__No preemption involved__:
+`Pending` -> `Allocated` -> `Deleted`
+
+__Preemption involved__:
+`Pending` -> `Preempting` -> `Allocated` -> `Being preempted` -> `Deleted`
+
+### Events
+
+__e<sub>c</sub>__:
+
+Condition: scheduler crashes and restarts.
+
+Operation: none.
+
+__e<sub>0</sub>__:
+
+Condition: all cells that the cell allocation algorithm decides to allocate to the AG are in `Free` or `Reserved` states (we assume the cell allocation algorithm has ensured that for every `Reserved` cell this AG's priority is higher than that of the AG currently associated with the cell, e.g., a `Preempting` AG).
+
+Operation:
+
+For all cells allocated to this AG:
+
+`Free` -> `Used` (by this AG) (e<sub>0</sub> in cell state machine);
+
+`Reserved` (by another AG) -> `Used` (by this AG) (e<sub>8</sub> in cell state machine).
+
+__e<sub>1</sub>__:
+
+Condition: there is at least one cell among those the algorithm decides to allocate to this AG in `Used` or `Reserving` states (we assume the cell allocation algorithm has ensured that for every `Used` or `Reserving` cell this AG's priority is higher than that of the AG currently associated with the cell).
+
+Operation:
+
+For all cells currently associated with other AGs:
+
+`Used` (by other AGs) -> `Reserving` (by this AG) (e<sub>2</sub> in cell state machine);
+
+`Reserving`/`Reserved` (by others AGs) -> `Reserving`/`Reserved` (by this AG) (e<sub>3</sub>/e<sub>6</sub> in cell state machine); 
```

Comment:
by other AGs

Reply:
Which one is by other AGs?

Reply:
I mean:
`Reserving` (by this AG) (e<sub>2</sub> 
->
`Reserving` (by this AGs) (e<sub>2</sub>